hugr-core/src/export.rs

```diff
@@ -767,31 +745,47 @@ impl<'a> Context<'a> {
         match t {
             SumType::Unit { size } => {
                 let items = self.bump.alloc_slice_fill_iter((0..*size).map(|_| {
-                    self.make_term(model::Term::List {
-                        items: &[],
-                        tail: None,
-                    })
+                    model::ListItem::Item(self.make_term(model::Term::List { items: &[] }))
                 }));
-                let list = model::Term::List { items, tail: None };
-                let variants = self.make_term(list);
+                let variants = self.make_term(model::Term::List { items });
                 self.make_term(model::Term::Adt { variants })
             }
             SumType::General { rows } => {
-                let items = self
-                    .bump
-                    .alloc_slice_fill_iter(rows.iter().map(|row| self.export_type_row(row)));
-                let list = model::Term::List { items, tail: None };
+                let items = self.bump.alloc_slice_fill_iter(
+                    rows.iter()
+                        .map(|row| model::ListItem::Item(self.export_type_row(row, None))),
+                );
+                let list = model::Term::List { items };
                 let variants = { self.make_term(list) };
                 self.make_term(model::Term::Adt { variants })
             }
         }
     }
 
-    pub fn export_type_row<RV: MaybeRV>(&mut self, t: &TypeRowBase<RV>) -> model::TermId {
-        let mut items = BumpVec::with_capacity_in(t.len(), self.bump);
-        items.extend(t.iter().map(|row| self.export_type(row)));
+    pub fn export_type_row<RV: MaybeRV>(
+        &mut self,
+        row: &TypeRowBase<RV>,
+        tail: Option<model::TermId>,
+    ) -> model::TermId {
+        let mut items = BumpVec::with_capacity_in(row.len() + tail.is_some() as usize, self.bump);
+
+        for t in row.iter() {
+            match t.as_type_enum() {
```

Comment:
Consider `items.push(match t.as_type_enum() { .... })` or even `items.push(if let TypeEnum::RowVar(var)=t.as_type_enum() {...} else {...})`

Reply:
A shame there is no `from_sized_iter_in` so you could create the BumpVec and fill it in one go!

Reply:
Does `self.bump.alloc_slice_fill_iter` do it?

Reply:
There's a few places where `Bump::alloc_slice_fill_iter` would in principle work. I've kept the `BumpVec` at the moment since I am leaving open the option that export may be fallible and return a `Result`, in which case it'd be easier to change the `BumpVec` option.

Also `Bump::alloc_slice_fill_iter` takes an `ExactSizeIterator` which we do not have in some cases.

Reply:
> Also Bump::alloc_slice_fill_iter takes an ExactSizeIterator which we do not have in some cases.

Ah ok - I wondered whether there was a problem something like that, and then when I saw that every Iterator has a `size_hint` I thought maybe it'd be ok anyway (i.e. sizing with capacity on a best-effort basis). If `alloc_slice_fill_iter` of a non-ExactSizeIterator doesn't compile...then never mind, BumpVec it'll have to be.